File: ui/game.py
import pygame
import sys
from game_engine.reactor import ReactorCore

pygame.init()
screen = pygame.display.set_mode((800, 600))
pygame.display.set_caption("Reactor 5")
font = pygame.font.SysFont(None, 30)

clock = pygame.time.Clock()
reactor = ReactorCore()

def draw_text(text, x, y, color=(255, 255, 255)):
    surface = font.render(text, True, color)
    screen.blit(surface, (x, y))

def render():
    screen.fill((0, 0, 0))

    if not any(m.name == "Sensor Malfunction" for m in reactor.active_malfunctions):
        draw_text(f"Coolant: {reactor.coolant}", 20, 20)
        draw_text(f"Heat: {reactor.heat}", 20, 60)
        draw_text(f"Pressure: {reactor.pressure}", 20, 100)


    draw_text("Malfunctions:", 20, 160)
    for idx, mal in enumerate(reactor.active_malfunctions):
        draw_text(f"{idx + 1}. {mal.name} (Ticks Left: {mal.duration})", 40, 190 + 30 * idx)

    draw_text("1 - cool, 2 - valve, 3 - power, 4 - sensors, 5 - lockdown", 20, 530)
    draw_text("C - cool down, P - release pressure", 20, 560)

    pygame.display.flip()

def game_over():
    if reactor.coolant <= 0:
        return "You ran out of coolant! Sorry, it will hurt."
    if reactor.heat >= 100:
        return "Reactor meltdown! Heat exceeded safe limits."
    if reactor.pressure >= 120:
        return "Reactor exploded! Pressure too high."
    if reactor.red_button:
        return "Self-destructing protocol activated!"
    return None

def game_over_screen(message):
    screen.fill((0, 0, 0))
    game_over_font = pygame.font.SysFont(None, 60)
    msg_surface = game_over_font.render("GAME OVER", True, (255, 0, 0))
    reason_surface = font.render(message, True, (255, 255, 255))
    instruction_surface = font.render("Press [R] to try again or [ESC] to quit", True, (180, 180, 180))

    screen.blit(msg_surface, (screen.get_width() // 2 - msg_surface.get_width() // 2, 200))
    screen.blit(reason_surface, (screen.get_width() // 2 - reason_surface.get_width() // 2, 300))
    screen.blit(instruction_surface, (screen.get_width() // 2 - instruction_surface.get_width() // 2, 350))

    pygame.display.flip()

    while True:
        for event in pygame.event.get():
            if event.type == pygame.QUIT:
                pygame.quit()
                sys.exit()
            elif event.type == pygame.KEYDOWN:
                if event.key == pygame.K_ESCAPE:
                    pygame.quit()
                    sys.exit()
                elif event.key == pygame.K_r:
                    return
        clock.tick(10)

def run_game():
    global reactor
    reactor = ReactorCore()

    tick_rate = 3  # seconds per tick
    last_tick_time = pygame.time.get_ticks()

    while True:
        current_time = pygame.time.get_ticks()
        if (current_time - last_tick_time) >= tick_rate * 1000:
            reactor.tick()
            last_tick_time = current_time

            reason = game_over()
            if reason:
                game_over_screen(reason)
                return

        for event in pygame.event.get():
            if event.type == pygame.QUIT:
                pygame.quit()
                sys.exit()

            if event.type == pygame.KEYDOWN:
                if event.key == pygame.K_ESCAPE:
                    pygame.quit()
                    sys.exit()
                elif event.key == pygame.K_1:
                    reactor.fix_coolant_leak()
                elif event.key == pygame.K_2:
                    reactor.unjam_valve()
                elif event.key == pygame.K_3:
                    reactor.reset_circuit()
                elif event.key == pygame.K_4:
                    reactor.calibrate_sensors()
                elif event.key == pygame.K_5:
                    reactor.override_lockdown()
                elif event.key == pygame.K_c:
                    if not any(m.name == "Control Lockout" for m in reactor.active_malfunctions):
                        reactor.cool_down()
                elif event.key == pygame.K_p:
                    if not any(m.name == "Control Lockout" for m in reactor.active_malfunctions):
                        reactor.release_pressure()
                elif event.key == pygame.K_z:
                    reactor.red_button = True
        render()
        clock.tick(30)

while True:
    run_game()

File: static/game.js
// Game state
let reactor = {
    heat: 0,
    pressure: 0,
    coolant: 100,
    timeAlive: 0,
    coolDownCooldown: 0,
    pressureReleaseCooldown: 0,
    activeMalfunctions: [],
    malfunction: false,
    valveJammed: false,
    sensorFault: false,
    controlsLocked: false,
    redButton: false
};

// Game settings
const TICK_RATE = 3000; // 3 seconds per tick
let gameInterval;
let isGameOver = false;

// UI Elements
const coolantBar = document.getElementById('coolant-bar');
const pressureBar = document.getElementById('pressure-bar');
const heatBar = document.getElementById('heat-bar');
const malfunctionLog = document.querySelector('#malfunction-log');

// Initialize game
function initGame() {
    reactor = {
        heat: 0,
        pressure: 0,
        coolant: 100,
        timeAlive: 0,
        coolDownCooldown: 0,
        pressureReleaseCooldown: 0,
        activeMalfunctions: [],
        malfunction: false,
        valveJammed: false,
        sensorFault: false,
        controlsLocked: false,
        redButton: false
    };
    isGameOver = false;
    updateUI();
    gameInterval = setInterval(gameTick, TICK_RATE);
}

// Game tick function
function gameTick() {
    if (isGameOver) return;

    reactor.timeAlive++;
    reactor.heat += 5;
    reactor.pressure += 7;

    if (reactor.coolDownCooldown > 0) reactor.coolDownCooldown--;
    if (reactor.pressureReleaseCooldown > 0) reactor.pressureReleaseCooldown--;

    // Process active malfunctions
    for (let i = reactor.activeMalfunctions.length - 1; i >= 0; i--) {
        const mal = reactor.activeMalfunctions[i];
        applyMalfunctionEffect(mal);
        mal.duration--;
        if (mal.duration <= 0) {
            fixMalfunction(mal.name);
            reactor.activeMalfunctions.splice(i, 1);
        }
    }

    // Random malfunction chance
    if (!reactor.activeMalfunctions.length && Math.random() < 0.2) {
        triggerRandomMalfunction();
    }

    updateUI();
    checkGameOver();
}

// Malfunction handling
function triggerRandomMalfunction() {
    const malfunctions = [
        { name: "Coolant Leak", duration: 4, effect: "coolantLeak" },
        { name: "Valve Jam", duration: 5, effect: "valveJam" },
        { name: "Power Surge", duration: 3, effect: "powerSurge" },
        { name: "Sensor Malfunction", duration: 4, effect: "sensorMalfunction" },
        { name: "Control Lockout", duration: 4, effect: "controlLockout" }
    ];

    const mal = malfunctions[Math.floor(Math.random() * malfunctions.length)];
    reactor.activeMalfunctions.push(mal);
    reactor.malfunction = true;
    addLogEntry(`WARNING: ${mal.name} detected!`, 'warning');
}

function applyMalfunctionEffect(mal) {
    switch (mal.effect) {
        case "coolantLeak":
            reactor.coolant = Math.max(0, reactor.coolant - 5);
            break;
        case "valveJam":
            reactor.valveJammed = true;
            break;
        case "powerSurge":
            reactor.heat += 10;
            break;
        case "sensorMalfunction":
            reactor.sensorFault = true;
            break;
        case "controlLockout":
            reactor.controlsLocked = true;
            break;
    }
}

// Action handlers
function handleAction(action) {
    if (isGameOver) return;

    switch (action) {
        case 'coolant':
            if (reactor.coolant > 0 && reactor.coolDownCooldown === 0 && !reactor.controlsLocked) {
                reactor.heat = Math.max(0, reactor.heat - 10);
                reactor.coolant -= 10;
                reactor.coolDownCooldown = 3;
                addLogEntry("Cooling system activated", 'info');
            }
            break;
        case 'pressure':
            if (!reactor.valveJammed && reactor.pressureReleaseCooldown === 0 && !reactor.controlsLocked) {
                reactor.pressure = Math.max(0, reactor.pressure - 15);
                reactor.pressureReleaseCooldown = 3;
                addLogEntry("Pressure release valve opened", 'info');
            }
            break;
        case 'fix-leak':
            if (!reactor.controlsLocked) {
                fixMalfunction("Coolant Leak");
            }
            break;
        case 'unjam-valve':
            if (!reactor.controlsLocked) {
                fixMalfunction("Valve Jam");
            }
            break;
        case 'fix-surge':
            if (!reactor.controlsLocked) {
                fixMalfunction("Power Surge");
            }
            break;
        case 'fix-sensor':
            if (!reactor.controlsLocked) {
                fixMalfunction("Sensor Malfunction");
            }
            break;
        case 'panic':
            reactor.redButton = true;
            checkGameOver();
            break;
    }
    updateUI();
}

function fixMalfunction(name) {
    const index = reactor.activeMalfunctions.findIndex(m => m.name === name);
    if (index === -1) return;

    reactor.activeMalfunctions.splice(index, 1);
    reactor.malfunction = false;

    switch (name) {
        case "Valve Jam":
            reactor.valveJammed = false;
            reactor.pressure = Math.max(0, reactor.pressure - 5);
            break;
        case "Power Surge":
            reactor.heat = Math.max(0, reactor.heat - 10);
            break;
        case "Sensor Malfunction":
            reactor.sensorFault = false;
            break;
        case "Control Lockout":
            reactor.controlsLocked = false;
            break;
    }

    addLogEntry(`${name} fixed successfully`, 'info');
}

// UI updates
function updateUI() {
    const bars = document.querySelectorAll('.status-indicator');

    if (reactor.sensorFault) {
        bars.forEach(bar => {
            bar.style.display = 'none';
        });
    } else {
        bars.forEach(bar => {
            bar.style.display = 'flex';
        });
    }

    // Update status bars
    coolantBar.style.width = `${reactor.coolant}%`;
    pressureBar.style.width = `${(reactor.pressure / 120) * 100}%`;
    heatBar.style.width = `${reactor.heat}%`;

    coolantBar.style.backgroundColor = '#00ff00';

    // Update status bar colors based on danger levels
    updateStatusBarColor(pressureBar, reactor.pressure, 100);
    updateStatusBarColor(heatBar, reactor.heat, 80);

    updateButtonStates();
}

function updateStatusBarColor(bar, value, threshold) {
    if (value >= threshold) {
        bar.style.backgroundColor = '#ff0000';
    } else if (value >= threshold * 0.7) {
        bar.style.backgroundColor = '#ffff00';
    } else {
        bar.style.backgroundColor = '#00ff00';
    }
}

function updateButtonStates() {
    const buttons = document.querySelectorAll('.control-button');

    buttons.forEach(button => {
        if (!button.classList.contains('always-active')) {
            if (reactor.controlsLocked ||
                (button.classList.contains('coolant') && reactor.coolDownCooldown) ||
                (button.classList.contains('pressure') && (reactor.pressureReleaseCooldown || reactor.valveJammed))) {
                button.disabled = true;
                button.style.opacity = '0.5';
                button.style.pointerEvents = 'none';
            } else {
                button.disabled = false;
                button.style.opacity = '1';
                button.style.pointerEvents = 'auto';
            }
        }
    });
}

function addLogEntry(message, type = 'info') {
    const entry = document.createElement('div');
    entry.className = `malfunction-entry ${type}`;
    entry.textContent = message;
    malfunctionLog.appendChild(entry);
    malfunctionLog.scrollTop = malfunctionLog.scrollHeight;
}

// Game over handling
function checkGameOver() {
    let gameOverMessage = null;

    if (reactor.coolant <= 0) {
        gameOverMessage = "You ran out of coolant! Sorry, it will hurt.";
    } else if (reactor.heat >= 100) {
        gameOverMessage = "Reactor meltdown! Heat exceeded safe limits.";
    } else if (reactor.pressure >= 120) {
        gameOverMessage = "Reactor exploded! Pressure too high.";
    } else if (reactor.redButton) {
        gameOverMessage = "Self-destructing protocol activated!";
    }

    if (gameOverMessage) {
        isGameOver = true;
        clearInterval(gameInterval);
        addLogEntry("GAME OVER: " + gameOverMessage, 'danger');
        
        // Save score to server
        fetch('/api/save_score', {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
            },
            body: JSON.stringify({
                survival_time: reactor.timeAlive * 3
            })
        });

        // Show game over overlay
        const overlay = document.getElementById('game-over-overlay');
        const messageElement = document.getElementById('game-over-message');
        messageElement.textContent = gameOverMessage;
        overlay.classList.remove('hidden');
    }
}

function restartGame() {
    const overlay = document.getElementById('game-over-overlay');
    overlay.classList.add('hidden');
    initGame();
}

function quitToMenu() {
    window.location.href = '/';
}

// Initialize game when page loads
window.onload = initGame; 